packages/react/src/React.js

```diff
@@ -48,37 +48,57 @@ import {
 import ReactSharedInternals from './ReactSharedInternals';
 import {enableStableConcurrentModeAPIs} from 'shared/ReactFeatureFlags';
 
-const React = {
-  Children: {
+// This uses a function constructor to enforce a monomorphic object. Using an object literal
+// along with the conditional branches causes this object to deoptimize frequently otherwise.
+function ReactObject() {
+  this.Children = {
     map,
     forEach,
     count,
     toArray,
     only,
-  },
+  };
 
-  createRef,
-  Component,
-  PureComponent,
+  this.createRef = createRef;
+  this.Component = Component;
+  this.PureComponent = PureComponent;
 
-  createContext,
-  forwardRef,
-  lazy,
-  memo,
+  this.createContext = createContext;
+  this.forwardRef = forwardRef;
+  this.lazy = lazy;
+  this.memo = memo;
 
-  Fragment: REACT_FRAGMENT_TYPE,
-  StrictMode: REACT_STRICT_MODE_TYPE,
-  Suspense: REACT_SUSPENSE_TYPE,
+  this.Fragment = REACT_FRAGMENT_TYPE;
+  this.StrictMode = REACT_STRICT_MODE_TYPE;
+  this.Suspense = REACT_SUSPENSE_TYPE;
 
-  createElement: __DEV__ ? createElementWithValidation : createElement,
-  cloneElement: __DEV__ ? cloneElementWithValidation : cloneElement,
-  createFactory: __DEV__ ? createFactoryWithValidation : createFactory,
-  isValidElement: isValidElement,
+  this.createElement = __DEV__ ? createElementWithValidation : createElement;
+  this.cloneElement = __DEV__ ? cloneElementWithValidation : cloneElement;
+  this.createFactory = __DEV__ ? createFactoryWithValidation : createFactory;
+  this.isValidElement = isValidElement;
 
-  version: ReactVersion,
+  this.version = ReactVersion;
 
-  __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED: ReactSharedInternals,
-};
+  this.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED = ReactSharedInternals;
+
+  this.ConcurrentMode = null;
+  this.Profiler = null;
+  this.unstable_ConcurrentMode = null;
+  this.unstable_Profiler = null;
+
+  this.useCallback = null;
+  this.useContext = null;
+  this.useEffect = null;
+  this.useImperativeMethods = null;
+  this.useLayoutEffect = null;
+  this.useMemo = null;
+  this.useMutationEffect = null;
+  this.useReducer = null;
+  this.useRef = null;
+  this.useState = null;
+}
+
+const React = new ReactObject();
 
 if (enableStableConcurrentModeAPIs) {
   React.ConcurrentMode = REACT_CONCURRENT_MODE_TYPE;
```

Comment:
Do these lines not affect anything? (Any reason we shouldn't move them and the hooks assignments into the function too?)